src/man/grep.1.ronn
grep(1) -- Object pattern searcher
==========================

## SYNOPSIS

`grep` [-iv] [-f <field>] <regex> [<obj>...]<br>
`grep` [-v] -e <expr> [<obj>...]

## DESCRIPTION

**grep** searches input objects, selecting those which match a condition and
outputs them. In the first form, the condition is specified as a regexp(3)
pattern and is matched against the string representation of the <field>
attribute, if specified, or the string representation of the object. In the
second form, the condition is evaluated by a Javascript expression specified
in <expr> where `x` corresponds to a given input object.

Input objects which are strings or blobs are split into lines of text,
and each line is treated as a separate input object.

If no <obj> arguments are supplied, **grep** reads objects from standard input. 

The following options are available:

  * `-i`:
    Case insensitive matching. By default, **grep** is case sensitive.

  * `-v`:
    Selected objects are those _not_ matching the specified condition.

  * <regex>:
    A Javascript regular expression in the regexp(3) format.

  * `-e` <expr>:
    A Javascript expression in `x`, evaluated with `x` bound to the current
    input object.

## DIAGNOSTICS

**grep** exits with `true` if there was at least one match. Otherwise it
exits with a non-`true` value.

## NOTE

It must be stressed that **grep** is an _object_ grep, not a file grep. In
other words, `grep '[Pp]igshell' /home/doc/*md` treats the arguments as string
objects, not files whose contents must be matched against the given pattern.

To grep files in the manner of UNIX grep, use the fgrep(1) script.

## EXAMPLES

Find GDrive files whose mime type field contains the string 'spreadsheet':

    $ ls /gdrive | grep -f mime spreadsheet

Find text within an HTML file:

    $ cat http://pigshell.com/sample/life-expectancy.html | grep -i india

Find countries with life expectancy < 50:

    $ cat http://pigshell.com/sample/life-expectancy.html | table2js foo country data | grep -e '+x.data < 50' | printf

## SEE ALSO

fgrep(1), regexp(3)


Changelog.md
Changelog
=========

0.6.2
-----
  * Minor bugfixes

0.6.1
-----

  * Firefox Linux stack overflows reported by @ngkabra, fixed by tuning
    _nstack_max way down to rock bottom.

0.6.0
-----

  * Licensing now GPL
  * Project hosting moved to github


src/cmd/grep.js
/*
 * Copyright (C) 2012-2014 by Coriolis Technologies Pvt Ltd.
 * This program is free software - see the file COPYING for license details.
 */

function Grep(opts) {
    var self = this;

    Grep.base.call(self, opts);
    self.matched = false;
}

inherit(Grep, Command);

Grep.prototype.usage = 'grep         -- object pattern searcher\n\n' +
    'Usage:\n' +
    '    grep [-iv] [-f <field>] <regex> [<obj>...]\n' +
    '    grep [-v] -e <expr> [<obj>...]\n' +
    '    grep -h | --help\n\n' +
    'Options:\n' +
    '    -h --help    Show this message.\n' +
    '    -f <field>   Specify field of object\n' +
    '    <regex>      Javascript regular expression\n' +
    '    -e <expr>    Javascript expression, e.g. "x.attr == 10"\n' +
    '    -v           Invert sense of expression\n' +
    '    -i           Case-insensitive match\n';

Grep.prototype.next = check_next(do_docopt(objargs(function() {
    var self = this,
        exp = self.docopts['-e'];

    if (self.inited === undefined) {
        self.inited = true;
        if (self.docopts['<regex>']) {
            var opts = self.docopts['-i'] ? 'i' : '',
                re = new RegExp(self.docopts['<regex>'], opts),
                field = self.docopts['-f'],
                getfield;
            if (field) {
                self.getfield = eval_getfield(field, undefined);
                if (isstring(self.getfield)) {
                    return self.exit(self.getfield);
                }
                self.filter = function(x) {
                    var n = self.getfield(x);
                    if (n !== undefined) {
                        return re.test(n.toString());
                    } else {
                        return false;
                    }
                };
            } else {
                self.filter = function(x) {
                    return re.test(x.toString());
                };
            }
        }
        if (self.docopts['-e']) {
            self.filter = eval_getexp('!!(' + exp + ')');
            if (isstring(self.filter)) {
                return self.exit(self.filter);
            }
        }
        if (self.docopts['-v']) {
            self.filter = function(f) { return function(item) { return !f(item); };}(self.filter);
        }
        self._linebuffer = [];
        self._partline = '';
    }
    return next();

    function next() {
        if (self._linebuffer.length) {
            return process_item(self._linebuffer.shift());
        }
        self.unext({}, cef(self, function(res) {
            update_line_buffer.call(self, res, function() {
                return next();
            });
        }));
    }

    function process_item(item) {
        if (item === null) {
            self.done = self.matched ? true : 'No match';
            return self.eof();
        }

        var m = false;
        try {
            if (self.filter(item)) {
                self.matched = m = true;
            }
        } catch(err) {
            self.errmsg('Caught error: ' + err.message);
        }
        if (m) {
            return self.output(item);
        }
        return soguard(self, next);
    }
})));


function update_line_buffer(res, cb) {
    var self = this;
    if (!(res instanceof Array)) {
        res = [res];
    }
    function makelines(str) {
        var lines = (self._partline + str).split('\n'),
            last = lines.pop(),
        lines = lines.map(function(m) { return m + '\n'; });
        self._partline = last;
        self._linebuffer = self._linebuffer.concat(lines);
    }
    async.forEachSeries(res, function(item, acb) {
        if (isstring(item)) {
            makelines(item);
        } else if (item instanceof Blob) {
            to('text', item, {}, function(err, res) {
                if (err) {
                    return self.exit(err);
                }
                makelines(res);
                return soguard(self, acb.bind(null, null));
            });
            return;
        } else {
            if (self._partline) {
                self._linebuffer.push(self._partline);
                self._partline = '';
            }
            self._linebuffer.push(item);
        }
        return soguard(self, acb.bind(null, null));
    },
    function(err) {
        return cb();
    });
}

Command.register("grep", Grep);
